src/java.base/share/classes/java/lang/foreign/MemoryLayout.java

```diff
@@ -351,40 +423,41 @@ default MethodHandle byteOffsetHandle(PathElement... elements) {
      *     layout path.
      * </ul>
      * <p>
-     * The final address accessed by the returned var handle can be computed as follows:
+     * A layout path consists of zero or more <em>path sections</em> delimited by
+     * <a href=#deref-path-elements>dereference path elements</a>. For each section of the layout path, the  final
+     * accessed address can be computed as follows:
      *
      * <blockquote><pre>{@code
      * address = base(segment) + offset
      * }</pre></blockquote>
      *
-     * Where {@code base(segment)} denotes a function that returns the physical base address of the accessed
-     * memory segment. For native segments, this function just returns the native segment's
-     * {@linkplain MemorySegment#address() address}. For heap segments, this function is more complex, as the address
-     * of heap segments is virtualized. The {@code offset} value can be expressed in the following form:
-     *
-     * <blockquote><pre>{@code
-     * offset = c_1 + c_2 + ... + c_m + (x_1 * s_1) + (x_2 * s_2) + ... + (x_n * s_n)
-     * }</pre></blockquote>
-     *
-     * where {@code x_1}, {@code x_2}, ... {@code x_n} are <em>dynamic</em> values provided as {@code long}
-     * arguments, whereas {@code c_1}, {@code c_2}, ... {@code c_m} are <em>static</em> offset constants
-     * and {@code s_1}, {@code s_2}, ... {@code s_n} are <em>static</em> stride constants which are derived from
-     * the layout path.
+     * Where {@code base(segment)} denotes a function that returns the physical base address of the accessed memory segment.
+     * For native segments, this function just returns the native segment's {@linkplain MemorySegment#address() address}.
+     * For heap segments, this function is more complex, as the address of heap segments is virtualized. The {@code offset}
+     * is obtained as if by a call to a {@linkplain #byteOffsetHandle(PathElement...) byte offset handle} constructed using
+     * the path elements pertaining to a particular path section. Where the arguments are taken from the dynamic
+     * {@code long} coordinates corresponding to that path section.
+     * <p>
+     * The accessed address must be <a href="MemorySegment.html#segment-alignment">aligned</a> according to the {@linkplain
+     * #byteAlignment() alignment constraint} of the root layout (this layout). Note that this can be more strict
+     * (but not less) than the alignment constraint of the selected value layout.
      * <p>
      * Additionally, the provided dynamic values must conform to bounds which are derived from the layout path, that is,
      * {@code 0 <= x_i < b_i}, where {@code 1 <= i <= n}, or {@link IndexOutOfBoundsException} is thrown.
      * <p>
-     * Multiple paths can be chained, with <a href=#deref-path-elements>dereference path elements</a>.
+     * Multiple path sections can be chained, with <a href=#deref-path-elements>dereference path elements</a>.
```

Comment:
I've uploaded a new version which attempts to improve this by merging the two paragraphs about offset computation. I'm not sure which version is better, though... Given the length of the text, I feel like we need _some_ kind of concept that refer to sections of a path which don't contain dereference path elements.

Reply:
I agree that the text is still not optimal. I'd like to think about it some more, and maybe I can come up with some more concrete suggestions.

Reply:
Looking at this some more, here are some things which, I believe, can be improved:

* The text is unnecessarily speaking about _addresses_. This forces us to define an abstract function to compute the base address of a segment (we need to, since heap segments have a virtualized base). In reality, all this is unnecessary - there's just segments and offsets, we don't really care about the address.
* Ignoring dereference handles, the behavior of the var handle is straightforward: we just derive an offset into the segment by calling `offsetHandle` (all the indices are passed and validated there). Then it's as if we called MS::get with the target (value) layout and provided the offset obtained as above. Perhaps, relying on MS::get is suspicious here, as we're on a more primitive level - but I think just saying "accessing a value whose type is the carrier of the target layout, at the offset computed as XYZ in the provided segment" is ok (and that is what happens). I don't think there is a need to provide formulas.
* I realize that we're hand-waving a bit for when the target layout is an AddressLayout - in that case the resulting segment has certain size characteristics, which depend on the target layout of the address - and we should say that.
* Ad for dereference path elements (assuming we want to keep them, which I'm less sure about), we could use the existing structure where first we explain what happens when there's no dereference path element. Then, we could explain what happens when there's dereference path elements by saying that we split the path into multiple "chunks" (this is similar to what you have done earlier), and that we compute the var handle for each chunk, and show how things are wired together (at this point we should have explained all the concepts we needed for this).